Functorize over concurrency monad, and add zmq-lwt implementation

andersfugmann opened the pull request:
This PR is only informational, and should not be merged. 
Please see #54 for more information.

The PR implements both zmq-lwt and zmq-async. 
The zmq-async implementation should be identical to the existing (although I removed sexp)
the zmq-lwt is based on rgrinberg's [lwt-zmq](https://github.com/hcarty/lwt-zmq/blob/master/src/lwt_zmq.ml).

I have included some oddities for the sake of discussion.

andersfugmann commented on zmq-lwt/src/deferred.ml:
```diff
@@ -0,0 +1,24 @@
+
+  let create fd = Lwt_unix.of_unix_file_descr fd
+  let syscall_exn fd f = Lwt_unix.wrap_syscall Lwt_unix.Read fd f
+  let ready_to t kind =
```
In the original implementation, this should just always return `Ready. 
https://github.com/hcarty/lwt-zmq/blob/master/src/lwt_zmq.ml#L21

andersfugmann commented on zmq-lwt/src/deferred.ml:
```diff
@@ -0,0 +1,24 @@
+  type t = Lwt_unix.file_descr
+
+  let create fd = Lwt_unix.of_unix_file_descr fd
+  let syscall_exn fd f = Lwt_unix.wrap_syscall Lwt_unix.Read fd f
```
In the original implementation, we always give the argument `Lwt_unix.Read` here. I would think it should depend on the operation given, i.e. if its a read or a write.
https://github.com/hcarty/lwt-zmq/blob/master/src/lwt_zmq.ml#L22

hcarty replied:
Based on what I've read, the `Lwt_unix.Read` argument is correct here.  I'll see if I can dig up the mailing list post(s) I originally used to figure that out.

andersfugmann commented:
In general (both implementations), reading from the documentation, we should always consult ZMQ_EVENTS, as the ability to read from the fd does not mean that there is data available, and we ZMQ_EVENTS must be consulted before using the read/write status of the fd.

hcarty commented:
@andersfugmann Thanks for working on this!

When I wrote the Lwt interface initially I spent a long while testing and reading through zeromq docs, mailing list posts and asking questions on IRC.  [This mess](https://github.com/hcarty/lwt-zmq/commit/b3e82b9b200d3fabf02780dae163bfb3e81aab55) is the result of that investigation and a lot of (unfotunately lost to a dead machine) testing.  Something simpler would be great but I'd be very careful before changing the logic.

Should I submit a PR with the current lwt-zmq implementation similar to what @rgrinberg did for his Async version?  Or should we skip that and focus on this PR?

rgrinberg commented:
I'd like to see @hcarty's original version included and seeing a commit where it is overwritten by the proposal here. So that we have a way to reference the change that occurred to the lwt backend due to this parameterization. It would be useful as a commit to refer to, and bisect with.

rgrinberg commented on zmq-deferred/src/jbuild:
```diff
@@ -0,0 +1,6 @@
+(library
+ ((name zmq_deferred)
+  (wrapped true)
+  (libraries (zmq base))))
```
Can't we drop base here? If we're able to do that, we might just be able to include this module in `zmq` itself. Also taking the opportunity to make it build with `wrapped true` as well. In general, I don't see this library as useful on its own, so I'd like to look for ways to avoid having it visible to the user.

P.S. this will build locally, but it will actually install a broken META file (dune bug) because `zmq_deferred` also needs a `public_name`. I assume you just omitted that by accident?

andersfugmann replied:
I've given zmq-deferred the publicname zmq.deferred, which makes it part of zmq

hcarty commented:
Current lwt-zmq import PR in as #57

andersfugmann commented on zmq-async/src/zmq_async.ml:
```diff
@@ -1,89 +1 @@
-open Sexplib.Conv
-
-module Socket = struct
-  exception Break_event_loop [@@deriving sexp_of]
```
I have removed the sexp stuff, as AFAICS it is not used anywhere in the zmq code. Removing it allows us to not depend on base for the zmq-lwt version

rgrinberg commented:
I see. That's acceptable for now. I will try to reintroduce it in a way
that doesn't add a needless dependency to lwt. Async users except sexp
conversions for all their functions. Especially exceptions as those get
formatted like sexps by monitors.

On Fri, Mar 9, 2018, 2:00 PM Anders Peter Fugmann <notifications@github.com>
wrote:

> *@andersfugmann* commented on this pull request.
> ------------------------------
>
> In zmq-async/src/zmq_async.ml
> <https://github.com/issuu/ocaml-zmq/pull/56#discussion_r173377539>:
>
> > @@ -1,89 +1 @@
> -open Base
> -open Async_kernel
> -open Async_unix
> -open Sexplib.Conv
> -
> -module Socket = struct
> -  exception Break_event_loop [@@deriving sexp_of]
>
> I have removed the sexp stuff, as AFAICS it is not used anywhere in the
> zmq code. Removing it allows us to not depend on base for the zmq-lwt
> version
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/issuu/ocaml-zmq/pull/56#pullrequestreview-102558558>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIe-9fkqBt5DA7RwfRrRNBMl4on6sJrks5tciijgaJpZM4SjQno>
> .
>

andersfugmann commented:
@hcarty I have toyed which this before, and the edge triggering on the FD is a bit strange. My example above is based on what the docs says (api.zeromq.org/master:zmq-getsockopt#ZMQ_FD), bit its completely untested. It would be great to have some tests to validate the implementation, and validate assumptions made on the ZMQ implementation

hcarty commented:
@andersfugmann Agreed re: tests.  The original tests I had were effectively throwing lots of messages around concurrently and running (potentially for hours) until it failed... not the most elegant!

Part of the challenge with zeromq's FD is that it's not, as I understand, the actual communication socket.  It's representative of zeromq's internal communication state.  IIRC that's why the code always waits for `Lwt_unix.Read`.  I'll try to dig up some of the extra articles, mailing list posts and IRC logs that lead to the lwt-zmq as it currently exists.  That may be a bit of a challenge though as that part of the code was written almost six years ago.

hcarty commented:
This writeup came later, but does cover some relevant material: https://funcptr.net/2012/09/10/zeromq---edge-triggered-notification/

It references the [documentation](http://api.zeromq.org/4-2:zmq-getsockopt) which states:
>The ZMQ_FD option shall retrieve the file descriptor associated with the specified socket. The returned file descriptor can be used to integrate the socket into an existing event loop; the ØMQ library shall signal any pending events on the socket in an edge-triggered fashion by making the file descriptor become ready for reading.

*ready for reading* is why `Lwt_unix.Read` is checked.

andersfugmann commented:
@hcarty Yes. you are correct that the fd should only be tested for reading. 
The way I read the documentation is that the fd has nothing to do with data, but only ZMQ_EVENTS (The section right above ZMQ_FD), when the fd is ready for reading a new event state may be present. Hence there is no need to use the syscall wrapper at all, and we should consult ZMQ_EVENTS before even considering waiting for the FD to become readable. 

I'm currently writing some tests to try to validate this assumption.

andersfugmann commented:
I rebased on current master so it now tracks the original lwt-zmq implementation. 
I have rewritten send/recv based on my understanding of the spec, which seems to be backed by my fd tests, although I have not stress-tested it.

andersfugmann commented:
Reading further on this, I thin we need a different approach. The state of a zmq-socket can also be updated by send and recv, so we need a queue of threads to wakeup.
Working on that atm

andersfugmann commented:
Now all sends and receives are placed on queues, and only one thread is responsible for sending and receiving based on the state of the socket. 
I've implemented the system in such a way that sends are prioritized over recv to make sure we keep the network busy. It should not matter much as all operations are non-blocking anyways. 

Still untested. Tests to come.

hcarty commented on zmq/src/jbuild:
```diff
@@ -4,6 +4,7 @@
  ((name zmq)
   (public_name zmq)
   (wrapped false)
+  (flags (:standard "-w" "-37"))
```
What triggers warning 37?

andersfugmann replied:
Lots of (with different kind of constructors). 

```File "zmq/src/ZMQ.ml", line 663, characters 0-514:
Error (warning 37): constructor I_ETERM is never used to build values.
(However, this constructor appears in patterns.)
```
I don't think ocaml can see that these are constructed in the C code.

hcarty replied:
Makes sense - thanks.

hcarty commented:
I'm ok with this being merged.

rgrinberg commented:
Seems like it would be better to disable the warnings at type definition
site. Or maybe even attach the warning removal to the offending
constructors themselves

On Wed, Mar 14, 2018, 5:47 AM Anders Peter Fugmann <notifications@github.com>
wrote:

> *@andersfugmann* commented on this pull request.
> ------------------------------
>
> In zmq/src/jbuild
> <https://github.com/issuu/ocaml-zmq/pull/56#discussion_r174295575>:
>
> > @@ -4,6 +4,7 @@
>   ((name zmq)
>    (public_name zmq)
>    (wrapped false)
> +  (flags (:standard "-w" "-37"))
>
> Lots of (with different kind of constructors).
>
> Error (warning 37): constructor I_ETERM is never used to build values.
> (However, this constructor appears in patterns.)
>
> I don't think ocaml can see that these are constructed in the C code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/issuu/ocaml-zmq/pull/56#discussion_r174295575>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIe-wqcEmjBVS9MMVVUOjKLFVMHmzazks5teD6MgaJpZM4SjQno>
> .
>

rgrinberg commented:
Yeah, this all looks very promising to me. A couple of points to consider:

* I'm not really a fan of using Async's deferred terminology for both async and lwt. I wonder if using a more neutral term like `Future` would be a little more clear.

* How about pulling out the `'a Make(T).t` outside of the functor. This will require giving it a couple of type parameters - one for the socket and the other for the condition variable. This way, we can share some functions that don't require a particular backend (string conversion for example). I think this is has a minor benefit for people trying to run backend independent functions.

andersfugmann commented:
I've disabled the warnings only where the constructors are defined.

andersfugmann commented:
I'm Ok with changing the name Deferred to something else. Futures is what Lwt calls it (And I'm really an async fan), and It makes me thing of Js. We could change it to `Concurrency` or `Monad` or `Effects` (in eager anticipation of algebraic effects).

hcarty commented:
FWIW, Lwt's new terminology uses "promises".

I like `Future` or `Concurrency` as they're more likely to be understood by a reader new to the project.

andersfugmann commented:
Wrt. moving some parts outside the functor I would need to see some examples of how this benefits. In any case I would like to wait untill this PR is merged. Its quite but and long lived already.

rgrinberg commented:
Concurrency is way too long so definitely future over that.

On Wed, Mar 21, 2018, 1:45 AM Hezekiah M. Carty <notifications@github.com>
wrote:

> FWIW, Lwt's new terminology uses "promises".
>
> I like Future or Concurrency as they're more likely to be understood by a
> reader new to the project.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/issuu/ocaml-zmq/pull/56#issuecomment-374692276>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIe-0A-y4Jt5RroTfGiDztrH32ycefSks5tgUBCgaJpZM4SjQno>
> .
>

andersfugmann commented:
Bikesheding. Lets keep it as is.

rgrinberg commented:
That's fine with me. The benefits are pretty small anyway.

On Wed, Mar 21, 2018, 1:46 AM Anders Peter Fugmann <notifications@github.com>
wrote:

> Wrt. moving some parts outside the functor I would need to see some
> examples of how this benefits. In any case I would like to wait untill this
> PR is merged. Its quite but and long lived already.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/issuu/ocaml-zmq/pull/56#issuecomment-374692580>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIe-xj81Gt3Bi9fTZutPmvstoG2Luufks5tgUBygaJpZM4SjQno>
> .
>

rgrinberg commented on zmq-async/src/jbuild:
```diff
@@ -4,5 +4,4 @@
  ((name zmq_async)
   (public_name zmq-async)
   (wrapped false)
```
`(wrapped false)` should be omitted here.

andersfugmann replied:
That would mean that Zmq_async should be referenced though
`Zmq_async.Zmq_async`? Or am I wrong?

If we don't wrap, we should make sure that all modules inside are prefixed with zmq_ to avoid clashes with other modules.

rgrinberg replied:
> That would mean that Zmq_async should be referenced though
Zmq_async.Zmq_async? Or am I wrong?

Not quite.

If you write a `zmq_async.ml` module, then this will be your entry level module for the library. All constituting modules will be accessible with their shortened names. For example, in the current setup, the `Deferred` module would be properly private and only available as `Zmq_async__Deferred` (as we'd like). If you'd like to make the Deferred module accessible to the user, you may include a `module Deferred = Deferred` line in your `Zmq_async` module. If you omit `Zmq_async.ml` then jbuilder will simply generate one for you with all the modules in your library aliases as:

```ocaml
module Deferred = Deferred (* the compilation unit actually Zmq_async__Deferred *)
module Foo = Foo
(* etc. *)
```

> If we don't wrap, we should make sure that all modules inside are prefixed with zmq_ to avoid clashes with other modules.

This is right. But note that `(wrapped false)` should only be used for single module libraries and only exists in jbuilder for compatibility.

andersfugmann replied:
Thats great -  I did not know that. 
Re-reading the paragraph on wrapped does actually say it, but I completely missed it previously. 
Ill change asap.

andersfugmann replied:
I have wrapped the modules (removed the line), but now I cannot access Deferred. I'm accessing it though its internal (wrapped) name but its not nice. I could expose it, but I really like the idea of hiding it from users. An alternative would be to copy deferred.ml (though copy stanza) into the test dir. Suggestions?

rgrinberg commented on zmq-deferred/src/jbuild:
```diff
@@ -0,0 +1,7 @@
+(library
+ ((name zmq_deferred)
+  (public_name zmq.deferred)
+  (wrapped true)
```
I'd omit here it for consistency. Or add `(wrapped true)` elsewhere where it's implicit.

andersfugmann replied:
Will do.

rgrinberg commented on zmq-async.opam:
```diff
@@ -12,9 +12,6 @@ build-test: [["jbuilder" "runtest" "-p" name "-j" jobs]]
   "zmq"
   "jbuilder" {build}
-  "ppx_sexp_conv" {build & >= "v0.9.0"}
-  "base" {>= "v0.9.0"}
```
Note that we still depend on `Base` (see `Base.Info.of_string` usage)

andersfugmann replied:
Done.

andersfugmann commented:
Thanks for the reviews. Ill squash merge and close

rgrinberg commented:
Yeah, there's no perfect solution for this yet. What you're doing is fine
and I've done it for cohttp myself. It's not future proof as the name
mangling is only an implemention detail.

Another approach I've done is to add a private module with module Deferred
= Deferred. The user should know not to meddle in Zmq_async.Private

On Wed, Mar 21, 2018, 2:28 AM Anders Peter Fugmann <notifications@github.com>
wrote:

> *@andersfugmann* commented on this pull request.
> ------------------------------
>
> In zmq-async/src/jbuild
> <https://github.com/issuu/ocaml-zmq/pull/56#discussion_r175877165>:
>
> > @@ -4,5 +4,4 @@
>   ((name zmq_async)
>    (public_name zmq-async)
>    (wrapped false)
>
> I have wrapped the modules (removed the line), but now I cannot access
> Deferred. I'm accessing it though its internal (wrapped) name but its not
> nice. I could expose it, but I really like the idea of hiding it from
> users. An alternative would be to copy deferred.ml (though copy stanza)
> into the test dir. Suggestions?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/issuu/ocaml-zmq/pull/56#discussion_r175877165>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIe-5xD3F0J2rWT5Se1MYuXThi_w6SWks5tgUpfgaJpZM4SjQno>
> .
>
